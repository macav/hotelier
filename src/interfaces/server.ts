export enum Status {
  RUNNING = 'running',
  RESTARTING = 'restarting',
  STOPPED = 'stopped',
  CRASHED = 'crashed',
}
export interface Server {
  id: string;
  status: Status;
  env: { [key: string]: any };
}
