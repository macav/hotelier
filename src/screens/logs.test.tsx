import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import Logs from './logs';
import fetchMock from 'jest-fetch-mock';
import { Status } from '../interfaces';

class MockLogRef {
  scrollTop = 50;
  clientHeight = 50;

  addEventListener() {
    return jest.fn();
  }

  get scrollHeight() {
    return 100;
  }
}
describe('Logs', () => {
  let rendered: renderer.ReactTestRenderer;
  let instance: Logs;
  const match = { params: { server: 'server1' } };

  beforeEach(() => {
    fetchMock.mockResponse(
      JSON.stringify({ server1: { status: Status.RUNNING } })
    );
    rendered = renderer.create(<Logs match={match as any} />, {
      createNodeMock: (element) => {
        if (element.type === 'pre') {
          return new MockLogRef();
        }
        return null;
      },
    });
    instance = (rendered.getInstance() as any) as Logs;
  });

  it('matches the snapshot', () => {
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('scrolls to bottom only if it is not at bottom', () => {
    instance.isAtBottom = false;
    instance.scrollToBottomIfNecessary();
    expect(instance.logsRef!.scrollTop).toEqual(50);
    instance.isAtBottom = true;
    instance.scrollToBottomIfNecessary();
    expect(instance.logsRef!.scrollTop).toEqual(100);
  });

  it('updates "isAtBottom" on scroll', () => {
    const spy = jest.spyOn(instance.logsRef!, 'scrollHeight', 'get');
    spy.mockReturnValue(10000);
    instance.onScroll();
    expect(instance.isAtBottom).toEqual(false);
    spy.mockRestore();
    instance.onScroll();
    expect(instance.isAtBottom).toEqual(true);
  });

  it('does not error when ref is not set', () => {
    instance.logsRef = null;
    expect(() => instance.onScroll()).not.toThrowError();
  });

  it('updates the state of logs', () => {
    (global as any).ipcRenderer = {
      on: (name: string, callback: any) =>
        callback(null, { id: 'server1', output: 'log output' }),
    };
    instance.watch();
    expect(instance.state.logs.server1.length).toEqual(1);
    expect(rendered.toJSON()).toMatchSnapshot('with logs');
  });

  it('clears the logs', () => {
    instance.setState({ logs: { server1: [{ id: '1', html: 'log' }] } });
    const wrapper = shallow(<Logs match={match as any} />);
    wrapper.find('button').first().simulate('click');
    expect((wrapper.instance() as Logs).state.logs.server1).toEqual([]);
  });
});
