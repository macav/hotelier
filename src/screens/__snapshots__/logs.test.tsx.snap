// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs matches the snapshot 1`] = `
<div
  className="h-100"
>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-primary sticky-top py-1"
  >
    <div
      className="mr-auto font-weight-bold text-center text-white"
    >
      server1
       logs
    </div>
    <button
      className="btn btn-primary restart-button"
      onClick={[Function]}
      title="Restart"
    >
      <span
        className="fas fa-redo "
      />
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      title="Clear logs"
    >
      <i
        className="fas fa-eraser"
      />
    </button>
  </nav>
  <pre
    className="logs-window"
  />
</div>
`;

exports[`Logs updates the state of logs: with logs 1`] = `
<div
  className="h-100"
>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-primary sticky-top py-1"
  >
    <div
      className="mr-auto font-weight-bold text-center text-white"
    >
      server1
       logs
    </div>
    <button
      className="btn btn-primary restart-button"
      onClick={[Function]}
      title="Restart"
    >
      <span
        className="fas fa-redo "
      />
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      title="Clear logs"
    >
      <i
        className="fas fa-eraser"
      />
    </button>
  </nav>
  <pre
    className="logs-window"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "log output",
        }
      }
    />
  </pre>
</div>
`;
